### PR TITLE
Added documentation for device controller

### LINKED ISSUE #742: Documentation for device controller missing
**What would you like to be added**:
Documentation for device controller is missing and needs to be added.

**Why is this needed**:
It would help users understand the implementation of device controller

### PR DESCRIPTION
**Pull request by sujithsimon22**
**What type of PR is this?**
 /kind documentation

**What this PR does / why we need it**:
This PR contains documentation for device controller which will help users understand its working

**Which issue(s) this PR fixes**:

Fixes #742

### CODE REVIEW

File: docs/modules/cloud/device_controller.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+|Device Node Membership Updated | The device controller sends a membership update event to the edge node.|
+|Device  Twin Desired State Updated | The device controller sends a twin update event to the edge node.|
+|Device Model Updated           |  TODO: What happens to existing devices using this model which are generating telemetry data as per the old model ? Do we update the config map associated with the devices which are using this device model ?|
+|Device Deleted                 | The controller sends the device twin delete event to delete all device twins associated with the device. It also deletes config maps associated with the device and this delete event is synced to the edge. The mapper application effectively stops operating on the device.|
+|Device Model Deleted           |  The controller needs to run [`finalizers`](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers) to ensure that all device instances using the device model are deleted first, and only then should the model deletion proceed.|
```
Comment (rohitsardesai83) [inline docs/modules/cloud/device_controller.md]: we need to remove this row as it is not supported yet. in the design proposal , its ok to mention.

File: docs/modules/cloud/device_controller.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+|-------------------------------|---------------------------------------------- |
+|New Device Model Created       |  NA                                           |
+|New Device Created             | The controller creates a new config map to store the device properties and visitors defined in the device model associated with the device.  This config map is stored in etcd. The existing config map sync mechanism in the edge controller is used to sync the config map to the egde. The mapper application running in a container can get the updated config map and use the property and visitor metadata to access the device. The device controller additionally reports the device twin metadata updates to the edge node.|
+|Device Node Membership Updated | The device controller sends a membership update event to the edge node.|
+|Device  Twin Desired State Updated | The device controller sends a twin update event to the edge node.|
+|Device Model Updated           |  TODO: What happens to existing devices using this model which are generating telemetry data as per the old model ? Do we update the config map associated with the devices which are using this device model ?|
```
Comment (rohitsardesai83) [inline docs/modules/cloud/device_controller.md]: remove this row as it is not clear. in the docs , we publish only what is available.

File: docs/modules/cloud/device_controller.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
+# Device Controller
+
+
+## Device Controller Overview
+ The device controller is the cloud component of KubeEdge which is responsible for device management. Device management im KubeEdge is implemented by making use of Kubernetes
+ [Custom Resource Definitions (CRDs)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) to describe device metadata/status and device controller to synchronize these device updates between edge and cloud. 
+ The device controller starts two separate goroutines called `upstream controller` and `downstream controller`. These are not separate controllers as such but named here for clarity.
+
+The device controller makes use of device model and device instance to implement device management :
+ - **Device Model**: A `device model` describes the device properties exposed by the device and property visitors to access these properties. A device model is like a reusable template using which many devices can be created and managed.
```
Comment (rohitsardesai83) [inline docs/modules/cloud/device_controller.md]: we can add links to api types in the design document.

Comment (rohitsardesai83) [reply docs/modules/cloud/device_controller.md]: also add samples for device models and instances which are available in the repository.

File: docs/modules/cloud/device_controller.md (hunk 2 above)
Comment (sujithsimon22) [reply docs/modules/cloud/device_controller.md]: Done, thanks

File: docs/modules/cloud/device_controller.md (hunk 3 above)
Comment (sujithsimon22) [reply docs/modules/cloud/device_controller.md]: Done, thanks

File: docs/modules/cloud/device_controller.md (hunk 1 above)
Comment (sujithsimon22) [reply docs/modules/cloud/device_controller.md]: Done, thanks

File: docs/modules/cloud/device_controller.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ Details on device model definition can be found [here](https://github.com/kubeedge/kubeedge/blob/master/docs/proposals/device-crd.md#device-model-type-definition).
+ - **Device Instance**: A `device` instance represents an actual device object. It is like an instantiation of the `device model` and references properties defined in the model. The device spec is static while the device status contains dynamically changing data like the desired state of a device property and the state reported by the device.
+ Details on device instance definition can be found [here](https://github.com/kubeedge/kubeedge/blob/master/docs/proposals/device-crd.md#device-instance-type-definition).
+
+ **Note**: Sample device model and device instance for a few protocols can be found at $GOPATH/src/github.com/kubeedge/kubeedge/build/crd-samples/devices
```
Comment (rohitsardesai83) [inline docs/modules/cloud/device_controller.md]: please give the complete url instead of $GOPATH style path.

**Comment by rohitsardesai83** [comment]
lets improve docs further in next iteration.

**Pull request merged**